### PR TITLE
feat(php): inheritance

### PR DESCRIPTION
**Pull request by dcb6**
This PR enables inheritance in PHP for types and inlined requests, which is achieved via PHP traits.

### CODE REVIEW

File: generators/php/model/src/generateTraits.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { ModelGeneratorContext } from "./ModelGeneratorContext";
+import { TraitGenerator } from "./trait/TraitGenerator";
+
+export function generateTraits(context: ModelGeneratorContext): void {
+    const extendedTypeIdsFromTypes: Set<string> = new Set(
```
Comment (amckinney) [inline generators/php/model/src/generateTraits.ts:5]: Easy lint fix here 
```suggestion
    const extendedTypeIdsFromTypes = new Set<string>(
```

File: generators/php/model/src/generateTraits.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            return endpoint.requestBody?.type === "inlinedRequestBody"
+                ? endpoint.requestBody.extends.map((declaredTypeName) => declaredTypeName.typeId)
+                : [];
+        });
+    for (const typeId of [...extendedTypeIdsFromTypes, ...extendedTypeIdsFromInlinedRequests]) {
+        const typeDeclaration = context.getTypeDeclarationOrThrow(typeId);
+        const objectTypeDeclaration = context.getUnderlyingObjectTypeDeclarationFromTypeDeclaration(typeDeclaration);
+        if (objectTypeDeclaration == null) {
+            throw new Error("Unexpected no object type declaration");
+        }
```
Comment (amckinney) [inline generators/php/model/src/generateTraits.ts:30]: nit: We could roll out a separate `getUnderlyingObjectTypeDeclarationOrThrow(typeId)` function to handle this block so the call-site reads a little simpler.

File: generators/php/codegen/src/ast/Class.ts [hunk 3]
```diff
@@ -20,6 +22,8 @@ export declare namespace Class {
         docs?: string;
         /* The class to inherit from if any */
         parentClassReference?: AstNode;
+        /* The traits that this class uses, if any */
+        usedTraits?: ClassReference[];
```
Comment (amckinney) [inline generators/php/codegen/src/ast/Class.ts:26]: nit: I would update this to just be `traits`, then the helper is just `addTrait`. We can assume it's used.

File: generators/php/codegen/src/ast/Trait.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+export declare namespace Trait {
+    interface Args {
+        /* The name of the PHP trait */
+        name: string;
+        /* The namespace of the PHP trait */
+        namespace: string;
+        /* Docs associated with the trait */
+        docs?: string;
+        /* The traits that this trait uses, if any */
+        usedTraits?: ClassReference[];
```
Comment (amckinney) [inline generators/php/codegen/src/ast/Trait.ts:18]: Same thing with `usedTraits` -> `traits` here.

File: generators/php/model/src/object/ObjectGenerator.ts [hunk 5]
```diff
@@ -23,24 +23,17 @@ export class ObjectGenerator extends FileGenerator<PhpFile, ModelCustomConfigSch
...
-                    name: this.context.getPropertyName(property.name.name),
-                    access: "public",
-                    docs: property.docs,
-                    attributes: this.context.phpAttributeMapper.convert({
-                        type: convertedType,
-                        property
-                    })
-                })
-            );
+            clazz.addField(this.toField({ property, inherited: false }));
```
Comment (amckinney) [inline generators/php/model/src/object/ObjectGenerator.ts:33]: nit: We can drop the `inherited: false` parameter if we make the parameter optional in `toField`.

**Review by amckinney — APPROVED** [review]
Nice, this looks great! It might be worth adding a lightweight test that we can generate into the SDK to confirm the traits are serialized like we expect.

File: generators/php/model/src/generateTraits.ts (hunk 1 above)
Comment (dcb6) [reply generators/php/model/src/generateTraits.ts]: ty! looking at this prompted to make a fix and wait to create the set around the concatenated final list of typeids 👍

File: generators/php/model/src/generateTraits.ts (hunk 2 above)
Comment (dcb6) [reply generators/php/model/src/generateTraits.ts]: i modified the existing methods to throw instead! let me know if you'd still like changes

i do need the `typeDeclaration` in the `TraitGenerator`, so the method still takes in a `typeDeclaration`

File: generators/php/codegen/src/ast/Trait.ts (hunk 4 above)
Comment (dcb6) [reply generators/php/codegen/src/ast/Trait.ts]: yeah good call, i wasn't sure if it was weird for a `Trait` field to have a `traits` field, but with your blessing i'm comfortable with it :)

File: generators/php/model/src/object/ObjectGenerator.ts (hunk 5 above)
Comment (dcb6) [reply generators/php/model/src/object/ObjectGenerator.ts]: nice, done!

**Pull request merged**